handler.go

```diff
@@ -26,64 +26,82 @@ var (
 	// only ever registered once.
 	delegateErrorHandlerOnce sync.Once
 
-	// Comiple time check that loggingErrorHandler implements ErrorHandler.
-	_ ErrorHandler = (*loggingErrorHandler)(nil)
+	// Compile-time check that delegator implements ErrorHandler.
+	_ ErrorHandler = (*delegator)(nil)
 )
 
-// loggingErrorHandler logs all errors to STDERR.
-type loggingErrorHandler struct {
+type holder struct {
+	eh ErrorHandler
+}
+
+func defaultErrorHandler() *atomic.Value {
+	v := &atomic.Value{}
+	v.Store(holder{eh: &delegator{l: log.New(os.Stderr, "", log.LstdFlags)}})
+	return v
+}
+
+// delegator logs errors if no delegate is set, otherwise they are delegated.
+type delegator struct {
 	delegate atomic.Value
 
 	l *log.Logger
 }
 
-// setDelegate sets the ErrorHandler delegate if one is not already set.
-func (h *loggingErrorHandler) setDelegate(d ErrorHandler) {
-	if h.delegate.Load() != nil {
-		// Delegate already registered
-		return
-	}
+// setDelegate sets the ErrorHandler delegate.
+func (h *delegator) setDelegate(d ErrorHandler) {
+	// It is critical this is guarded with delegateErrorHandlerOnce, if it is
+	// called again with a different concrete type it will panic.
 	h.delegate.Store(d)
 }
 
-// Handle implements ErrorHandler.
-func (h *loggingErrorHandler) Handle(err error) {
+// Handle logs err if no delegate is set, otherwise it is delegated.
+func (h *delegator) Handle(err error) {
 	if d := h.delegate.Load(); d != nil {
 		d.(ErrorHandler).Handle(err)
 		return
 	}
 	h.l.Print(err)
 }
 
-// GetErrorHandler returns the global ErrorHandler instance. If no ErrorHandler
-// instance has been set (`SetErrorHandler`), the default ErrorHandler which
-// logs errors to STDERR is returned.
+// GetErrorHandler returns the global ErrorHandler instance.
+//
+// The default ErrorHandler instance returned will log all errors to STDERR
+// until an override ErrorHandler is set with SetErrorHandler. All
+// ErrorHandler returned prior to this will automatically forward errors to
+// the set instance instead of logging.
+//
+// Subsequent calls to SetErrorHandler after the first will not forward errors
+// to the new ErrorHandler for prior returned instances.
 func GetErrorHandler() ErrorHandler {
-	return globalErrorHandler
+	return globalErrorHandler.Load().(holder).eh
 }
 
-// SetErrorHandler sets the global ErrorHandler to be h.
+// SetErrorHandler sets the global ErrorHandler to h.
+//
+// The first time this is called all ErrorHandler previously returned from
+// GetErrorHandler will send errors to h instead of the default logging
+// ErrorHandler. Subsequent calls will set the global ErrorHandler, but not
+// delegate errors to h.
 func SetErrorHandler(h ErrorHandler) {
 	delegateErrorHandlerOnce.Do(func() {
 		current := GetErrorHandler()
 		if current == h {
 			return
 		}
-		if internalHandler, ok := current.(*loggingErrorHandler); ok {
+		if internalHandler, ok := current.(*delegator); ok {
 			internalHandler.setDelegate(h)
 		}
 	})
```

Comment:
I suggest using a global delegator, `globalDelegator`, instead of fetching a delegator instance from `GetErrorHandler` is a better approach. 

Hence, the rules of `GetErrorHandler()` become more intuitive: users get `default ErrorHandler`(Delegator) before the first call to `SetErrorHandler` and get `new ErrorHandler` after calling `SetErrorHandler`; `default ErrorHandler` always uses the latest `ErrorHandler`.

We can also get rid of `delegateErrorHandlerOnce` in this case.

```go
func SetErrorHandler(h ErrorHandler) {
	current := GetErrorHandler()
	if current == h {
		return
	}
	globalDelegator.setDelegate(h)
	globalErrorHandler.Store(holder{eh: h})
}
```

Updated `TestDelegatedHandler`

```go
func (s *HandlerTestSuite) TestDelegatedHandler() {
	eh := GetErrorHandler()

	newErrLogger := new(errLogger)
	newErrLogger2 := new(errLogger)
	SetErrorHandler(&logger{l: log.New(newErrLogger, "", 0)})
	SetErrorHandler(&logger{l: log.New(newErrLogger2, "", 0)})

	errs := []string{"TestDelegatedHandler"}
	eh.Handle(errors.New(errs[0]))
	s.Assert().Equal(errs, newErrLogger2.Got())
}
```

Reply:
This approach will set the delegate multiple times, something not done for any of the other global variables and an action that the workflow was intentionally designed to not do.

If there is a desire to switch this algorithm it seems like it would warrant an issue to discuss the topic on its own as it should require changing all global variable behavior. I do not want to make this change in this PR.